File: src/10.c
#include "util.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Coord {
    int row;
    int col;
} Coord;

typedef struct Loop {
    Coord curr;
    Coord prev;
    char curr_type;
    Coord *vertices;
    int length;
    int capacity;
} Loop;

void append_vertex(Loop *loop) {
    ++loop->length;
    if (loop->length == loop->capacity) {
        loop->capacity *= 2;
        int new_size = sizeof(Coord) * loop->capacity;
        loop->vertices = realloc(loop->vertices, new_size);
    }
    loop->vertices[loop->length - 1] = loop->curr;
}

Coord find_start(char **diagram) {
    Coord start = {-1, -1};
    for (int i = 0; diagram[i]; ++i) {
        char *s = strchr(diagram[i], 'S');
        if (s) {
            start.row = i;
            start.col = s - diagram[i];
            break;
        }
    }
    return start;
}

void look(Loop *loop, Coord dir, char *connects, char **diagram) {
    Coord next = {loop->curr.row + dir.row, loop->curr.col + dir.col};
    Coord prev = loop->prev;

    char *can_move = strchr(connects, loop->curr_type);
    int backtracking = prev.row == next.row && prev.col == next.col;

    if (can_move && !backtracking) {
        loop->prev = loop->curr;
        loop->curr = next;
        loop->curr_type = diagram[next.row][next.col];
        append_vertex(loop);
    }
}

void traverse(Loop *loop, char **diagram) {
    char *directions[] = {"SJL|", "SF7|", "SFL-", "S7J-"};
    Coord coords[] = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};

    char curr_type = '\0';
    while (curr_type != 'S') {
        for (int i = 0; i < 4; ++i) {
            look(loop, coords[i], directions[i], diagram);
        }
        curr_type = loop->curr_type;
    }
}

int get_interior_tiles(Loop *loop) {
    int sum = 0;
    for (int i = 0; i < loop->length; ++i) {
        int prev_idx = i - 1 < 0 ? loop->length - 1 : i - 1;
        Coord p_0 = loop->vertices[i];
        Coord p_1 = loop->vertices[prev_idx];
        sum += (p_1.col * p_0.row) - (p_0.col * p_1.row);
    }
    return (sum - loop->length) / 2 + 1;
}

int main(void) {

    char **diagram = file_to_str_array("./input/10");
    Coord start = find_start(diagram);

    Loop loop = {.curr = start, .curr_type = 'S', .length = 0, .capacity = 7};
    loop.vertices = calloc(sizeof(Coord), loop.capacity);
    loop.vertices[loop.length] = loop.curr;

    traverse(&loop, diagram);
    int farthest = loop.length / 2 + loop.length % 2;
    int interior = get_interior_tiles(&loop);

    printf("\nAoC 2023 Day 10\n");
    printf("Loop starts at: %d %d\n", start.row, start.col);
    printf("Loop length: %d\n", loop.length);
    printf("Part 1: Farthest pipe is %d pipes away\n", farthest);
    printf("Part 2: There are %d interior tiles\n", interior);
}

File: src/10.py
from attrs import define, field

INPUT = "./input/10"

connects: dict[tuple[int, int], str] = {
    (-1, 0): "SJL|",
    (1, 0): "SF7|",
    (0, -1): "S7J-",
    (0, 1): "SFL-",
}


@define
class Pipe:
    row: int = field()
    col: int = field()
    rows: list[str] = field()
    prev: tuple[int, int] = row, col
    char: str = "S"
    length: int = 0
    coords: list[tuple[int, int]] = []

    def get_interior_tiles(self):
        sum = 0
        for i in range(0, self.length):
            x_0, y_0 = self.coords[i]
            x_1, y_1 = self.coords[i - 1]
            sum += (x_0 * y_1) - (x_1 * y_0)
        return (sum - self.length) / 2 + 1

    def get_farthest_pipe(self):
        return self.length // 2 + self.length % 2


def look(pipe: Pipe, dir: tuple[int, int]) -> tuple[int, int]:
    row_next, col_next = pipe.row + dir[0], pipe.col + dir[1]
    can_move: bool = pipe.char in connects[dir]

    if can_move and pipe.prev != (row_next, col_next):
        pipe.prev = (pipe.row, pipe.col)
        pipe.coords.append(pipe.prev)
        pipe.char = pipe.rows[row_next][col_next]
        pipe.row = row_next
        pipe.col = col_next
        pipe.length += 1
        return pipe.row, pipe.col
    return -1, -1


def get_diagram(file: str) -> list[str]:
    with open(file, "r") as f:
        return [line.strip() for line in f.readlines()]


def find_start(diagram: list[str]) -> tuple[int, int]:
    row, col = 0, 0
    for line in diagram:
        col = line.find("S")
        if col >= 0:
            return row, col
        row += 1
    return -1, -1


def traverse(pipe: Pipe):
    pipe_type = ""
    while pipe_type != "S":
        look(pipe, (-1, 0))
        look(pipe, (1, 0))
        look(pipe, (0, -1))
        look(pipe, (0, 1))
        pipe_type = pipe.char


def main():
    diagram = get_diagram(INPUT)
    row, col = find_start(diagram)
    pipe = Pipe(row, col, diagram)
    traverse(pipe)

    print("\nAoC 2023 Day 10 - Part 1\n")
    print(f"Pipe length: {pipe.length}")
    print(f"Part 1: Farthest pipe is {pipe.get_farthest_pipe()} pipes away")
    print(f"Part 2: There are {pipe.get_interior_tiles()} interior tiles")


if __name__ == "__main__":
    main()
